README.md
slappy
======

An intelligent IRC bot that is capable of doing basic functions. Written in PHP.

> Basic commands.

This is the list that it currently can handle:
- server : shows the server slappy currently is connected;
- join : joins a channel;
- part : parts a channel;
- action : does a action like " *slappy is happy " ;
- do 
- bindings : shows the current bindings for the bot;
- quit : disconnects the bot from the irc;
- die : kills the bot ;


> Oper module (as of version 1.1);

This module can oper the bot on your irc server.
For the config parameters see oper.php in modules/ .

> NickServ moule

Yes, Slappy comes with a nickserv module that can help you to
identify the bot to the service. For more details see nickserv.php

> Installing

To install Slappy you need php installed. Version 4.3.1+ is recomended.
Clone this repo to your favourite directory. Then you need to create the config file or just edit slappy-config.sample and rename it to config.php.
After that you just need to run "php slappy.php" . If there are no errors then you've done everything right.

> Licence

See licence details in the LICENCE file.

> What can Slappy do ?

Public/private commands for slappy.
Use /msg <#chan,botnick> <cmd> or /notice <botnick> <cmd>
Currently support the commands: server, join, part, action, do,
bindings, rehash (not done yet), quit, die, mysql (not ready yet).


CHANGELOG
This is the changelog for slappy.


4.10.2004.v[1.1]

-created modules/ , where the core modules can be found;
-added the oper module, where you can oper the bot if you want;

3.10.2004.v[1.0]

-Initial release of slappy,wasn't not released for the public.


modules/api.php
<?
/*

   Slappy - a php ircbot
    Copyright (C) 2004-2005

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */
/*
This is the API for Slappy. It provides more easy-to-use functions
to the end-user.
*/

function msg($target,$text) {
    dump("PRIVMSG $target :$text");
}

function notice($target,$text) {
    dump("NOTICE $target :$text");
}

function jchan($chan) { // can be used as jchan($chan,$key) too
    $chan_and_key = implode(" ",func_get_args());
    dump("JOIN :$chan_and_key");
}

function part($chan) {
    dump("PART $chan");
}

function kick($chan,$nick) { // can be used as kick($chan,$nick,$reason) too
    $reason = '';
    if (func_num_args()>2) $reason = func_get_arg(2); 
    dump("KICK $chan $nick :$reason");
}

function mode($target,$modes) { // can be with more than two params too
    $args = func_get_args();
    $target = array_shift($args);
    $modes = array_shift($args);
    $params = implode(" ",$args);
    dump("MODE $target $modes :$params");
}

function topic($chan,$topic) { 
    dump("TOPIC $chan :$topic");
}

function quit() { //can be used as quit($reason) too
    $reason = '';
    if (func_num_args()) $reason = func_get_arg(0);
    dump("QUIT :$reason");
}
	
function action($target,$text) {
    dump("PRIVMSG $target :\001ACTION $text\001");
}

function version($target) {
    dump("PRIVMSG $target :\001VERSION\001");
}

function ping($target) {
    dump("PRIVMSG $target :\001PING ".time()."\001");
}

function wall($text) {
    dump("WALLOPS :$text");
}

function operwl($text) {
    dump("OPERWALL :$text");
}

function locops($text) {
    dump("LOCOPS :$text");
}

function get_user_info($user) {
    ereg("(.*)!(.*)@(.*)",$user,$regs);
    return $regs;
}

function get_nick($user) {
    $result = get_user_info($user);
    return $result[1];
}

function get_ident($user) {
    $result = get_user_info($user);
    return $result[2];
}

function get_host($user) {
    $result = get_user_info($user);
    return $result[3];
}

echo ">> Api module loaded\n";

?>


modules/raw.php
<?
/*

   Slappy - a php ircbot
    Copyright (C) 2004-2005

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */
// raw messages support.

bind('NickInUse','raw433');
function NickInUse() {
	bnick(rand_name());
}

bind('raw_002_handle','raw002');
function raw_002_handle($user, $params) {
	 ircserver('set',$user);
	 return;
}

function ircserver() {
    global $ircserver;
    if ((func_num_args() == 2) && (func_get_arg(0) == 'set')) {
    	$ircserver = func_get_arg(1);
    }
    return $ircserver;
}

bind('WhoisHandle','raw311');
function WhoisHandle($user,$params) {
	array_shift($params); // this is bot's nick, not useful 
	$arr['nick'] = array_shift($params);
	$arr['ident'] = array_shift($params);
	$arr['host'] = array_shift($params);
	array_shift($params); // useless *
	$arr['ircname'] = strip_colon_and_join($params);
	return $arr;
}

echo ">> Raw module loaded \n";

?>


modules/oper.php
<?
/*

   Slappy - a php ircbot
    Copyright (C) 2004-2005

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */
/*
This script provides oper support for slappy.
*/

$nick = "slappy"; // nick in the o:line
$pass = "mypass"; // password for /oper
$omodes = "+cks"; // modes to set after oper identification
$debug = "on"; // debug on/off
$debugchan = "#debug";

bind('operid',"PRIVMSG");

function operid($user,$whom,$msg) {
    global $nick, $pass, $debugchan;
    if (is_admin($user) == true) {
	if ($msg == bnick()." oper") {
	    if ($debug == "on") {
		msg($debugchan,"trying to oper...");
	    }
	    dump("oper $nick $pass");
	}
    }
}

bind('operok','raw381');

function operok($user,$params) {
    global  $debugchan, $omodes;
    if ($user == ircserver()) {
	$smsg = array();
        $smsg = explode(" ",$params);
	if ($debug == "on") {
	    msg($debugchan,"operation successful!");
	}
        dump("MODE ".bnick()." $omodes");
    }
}

bind('opernotok','raw491');

function opernotok($user,$params) {
    global $debugchan;
    if ($debug == "on") {
	msg($debugchan,"operation failed!;(");
    }
}

echo ">> Oper module loaded\n";

?>


modules/pubcmd.php
<?
/*

   Slappy - a php ircbot
    Copyright (C) 2004-2005

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */
/*
Public/private commands for slappy.
Use /msg <#chan,botnick> <cmd> or /notice <botnick> <cmd>
Currently support the commands: server, join, part, action, do,
bindings, rehash (not done yet), quit, die, mysql (not ready yet).

*/

bind('pubcmd',"PRIVMSG","NOTICE");

function pubcmd($user, $whom, $msg) {
    global $bnick, $version;
    $params = explode(" ",$msg);
    if (ereg("^#",$whom)) { // if they're calling the bot in a channel
	$mecalled = array_shift($params); // getting bots nick out
    } else { // private msg to the bot
	$mecalled = $bnick;
        $whom = get_nick($user);
    }
    $command = strtoupper(array_shift($params));
    if ($mecalled == $bnick) {
    	switch($command) {
	    case "SERVER":
		msg($whom,"I'm using ".ircserver());
		break;
	    case "\001PING":
		dump("NOTICE $whom :\001PING ".implode(" ",$params));
		break;
	    case "\001VERSION\001":
		dump("NOTICE ".get_nick($user)." :\001VERSION $version \001");
		break;
	    default:
		if (is_admin($user)) {
		    switch ($command) {
			case "JOIN":
			    call_user_func_array('jchan',$params);
			    break;
			case "PART":
			    part($params[0]);
			    break;
			case "ACTION":
			    action($whom,join(" ",$params));
			    break;
			case "DO":
			    dump(join(" ",$params));
			    break;
			case "BINDINGS":
			    msg($whom,"BINDINGS for $params[0] -> (".join(",",managebindings('list',$params[0])).")");
			    break;
			case "REHASH":
			    msg($whom,"rehashing...");
			    //rehash();
			    break;
			case "QUIT":
    			    irclog("cmd","ressurect requested by ".get_nick($user));
			    quit("requested by ".get_nick($user));
			    break;
			case "DIE":
			    quit("requested by ".get_nick($user));
			    irclog("cmd","die request by ".get_nick($user));
			    exit();
			    break;
			case "MYSQL":
			    $query = implode(" ",$params);
			    jmysql($query);
			    break;
			default:
			    msg($whom,"Not implemented... yet!");
			    break;
		    }
		}
		break;
        }
    }
}

echo ">> Public commands module loaded\n";
?>


modules/nickserv.php
<?
/*

   Slappy - a php ircbot
    Copyright (C) 2004-2005

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

$nspass = "mypass";
$debug = "on"; // debug on/off
$debugchan = "#debug";

bind('nsident',"NOTICE");

function nsident($user,$whom,$msg) {
    global $nspass, $debugchan;
    if (get_nick($user) == "NS") {
	if ("This nickname is owned by someone else" == $msg) {
	    msg("NS","identify $nspass");
	    irclog("NS","Trying to identify myself");
	    msg($debugchan,"trying to id");
	} elseif ("Password accepted - you are now recognized" == $msg) {
	    irclog("NS","Identification successful, yuppie!");
	    if ($debug == "on") {
    		msg($debugchan,"Identification successful: yuppie!");
	    }
	    msg("CS","op all");
	} elseif ("Password Incorrect" == $msg) {
	    irclog("NS","Identification failed: wrong password");
	    if ($debug == "on") {
		msg($debugchan,"Identification failed: wrong password");
	    }
	}
    }
}

echo ">> Nickserv module loaded\n";

?>


slappy.php
<?
/*

   Slappy - a php ircbot
    Copyright (C) 2004-2005

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */
if (!@include("config.php")) {
    die("The configuration file was not found!\n");
}

$version = "Slappy , a php irc bot";

$buser = "$bident";
$buser .= " ".rand_name();
$buser .= " ".rand_name();
$buser .= " :$brealname";

set_time_limit(0); // setting max exec time to zero

connect($ircserver);

function connect($ircserver) {
    global $socket,$ircserver,$bnick,$buser,$bchans,$autojoin;
    $socket = socket_create(AF_INET,SOCK_STREAM,SOL_TCP); // create the socket
    $connection = socket_connect($socket,$ircserver,6667); // connect

    //var_dump($connection); // THIS IS FOR DEBUG

    socket_write($socket,"USER $buser\r\n"); // send the Username
    socket_write($socket,"NICK $bnick\r\n"); // change our nickname
    
    if ($autojoin == "on") {
	foreach ($bchans as $bchan) {
	    socket_write($socket,"JOIN $bchan\r\n"); // join the channel
	
	}
    }
    ircloop();
}

function ircloop() {
    global $socket,$ircserver;
// read whatever IRC is telling us
    $datum="initial";
    $lastpiece="";
//sometimesh socket_read return "" when the connection is closed
    while ((!socket_last_error()) && ($datum != "")) {
//safe for Windows (PHP_NORMAL_READ doesn't work)
	$datum = $lastpiece.socket_read($socket,1024,PHP_BINARY_READ); // 255 is arbitrary
// preventing too much data from causing buffer overflow by spliting it
// and then agglutinating it to the next piece of data to make ot one
        if (ereg(".*\n(.+)",$datum,$regdata) && (!ereg("\n$",$regdata[1]))) {
    	    $lastpiece = $regdata[1];                                           
    	    $datum = substr($datum,0,strlen($datum)-strlen($lastpiece));
	} else {
    	    $lastpiece = "";
	}

// separating every line from the server
	foreach (explode("\r\n",chop($datum)) as $data) {
// debug
	    echo "data - $data\n";//<br>";
// debug unix time
	    echo time()."\n";//<br>";
// separating the data into an array
    	    $split = explode(" ",$data);
// cutting : and getting the nickname as $user
    	    $user = str_replace(":","",array_shift($split)); 
// getting the command
    	    $cmd = array_shift($split);
// taking the rest as parameters of the command
    	    $params = $split;
// the PING syntax is a little bit strange, so we're taking care of it
    	    if ($user == "PING") {
    		$user = ereg_replace("^:","",$cmd);
        	$cmd = "PING";
    	    }
	    respond($user, $cmd, $params);
	}
    }
    
    global $reconnect; // this should be at the top of the script
    if ($reconnect == "on") {
	connect($ircserver);
    } else {
	exit();
    }
}

function respond($user, $cmd, $params) {
    global $socket,$bnick,$nspass;
// uppering the cmd due RFC standarts
    if (strlen($cmd) == 3) { // if $cmd is raw numeric
    	mfunc("raw".$cmd,$user,$params); //send to our master function for event handling
    	// .inc scriptrs should use bind('function_name',"raw301") (for example)
    	// and then function function_name($user,$msg) 
    	// usually $user is the server sending messages
    	// raw001 will save the zeroes (otherwise 001 will become 1, later in bindings) 
    }
    $cmd = strtoupper($cmd);
    switch ($cmd) {
// event handling - very important
	case "PRIVMSG":
    	    $whom = array_shift($params); // where to (chan or bot)
            $params[0] = ereg_replace("^:","",$params[0]); // taking off the : sign from params
            $msg = implode(" ",$params);
            mfunc("PRIVMSG",$user,$whom,$msg);
	    if ($whom[0] = "#") {
		$chan = $whom;
		mfunc("chan",$user,$chan,$msg);
	    } else {
		mfunc("query",$user,$msg);
	    }
	    break;            
// we're done with the parsing of PRIVMSG
// the include() scripts now can do their job
// $whom - called in chan or bot
// $mecalled - channel or nickname - where is invoked
// $user - nick!ident@host
// $command - the command that was sent to us
        case "PING":
// playing PING-PONG with the server
	    $pingserver = $user;
	    mfunc("PING",$pingserver);
	    ircwrite("raw","PONG $user");
            break;
        case "JOIN":
	    $params[0] = ereg_replace("^:","",$params[0]);
	    $chan = $params[0];
	    mfunc("JOIN",$user,$chan);
            break;
        case "NOTICE":
	    $whom = array_shift($params);
            $params[0] = ereg_replace("^:","",$params[0]);
	    $msg = implode(" ",$params);
	    mfunc("NOTICE",$user,$whom,$msg);
            break;
        case "PART":
	    $params[0] = ereg_replace("^:","",$params[0]);
	    $chan = $params[0];
	    mfunc("PART",$user,$chan);
            break;
        case "NICK":
	    $params[0] = ereg_replace("^:","",$params[0]);
	    $newnick = $params[0];
	    mfunc("NICK",$user,$newnick);
            break;
        case "MODE":
	    $target = array_shift($params);
	    $modes = array_shift($params);
	    mfunc("MODE",$user,$target,$modes,$params);
            break;
        case "TOPIC":
	    $chan = array_shift($params);
	    $params[0] = ereg_replace("^:","",$params[0]);
	    $topic = implode(" ",$params);
	    mfunc("TOPIC",$user,$chan,$topic);
            break;
        case "KICK":
	    $chan = array_shift($params);
	    $knick = array_shift($params);
	    $params[0] = ereg_replace("^:","",$params[0]);
	    $reason = implode(" ",$params);
	    mfunc("KICK",$user,$chan,$knick,$reason);
            break;
        case "QUIT":
	    $params[0] = ereg_replace("^:","",$params[0]);
	    $reason = implode(" ",$params);
	    mfunc("QUIT",$user,$reason);
            break;
	case "WALLOPS":
	    $params[0] = ereg_replace("^:","",$params[0]);
	    $wallmsg = implode(" ",$params);
	    mfunc("WALLOPS",$user,$wallmsg);
	    break;
    }
}

function mfunc() {
    $args = func_get_args();
    $event = array_shift($args);
    foreach (listbindings($event) as $func) {
	    call_user_func_array($func,$args);
    }
}

function bind($function_name, $events) { // it can be called with more events, but at least one is compulsary
    $arr = func_get_args();
    $function = array_shift($arr);
    foreach ($arr as $event) {
	managebindings('set',$event,$function);
    }
}

function unbind($function_name, $events) { // it can be called with more events, but at least one is compulsary
    $arr = func_get_args();
    $function = array_shift($arr);
    foreach ($arr as $event) {
	managebindings('unset',$event,$function);
    }
}
						
function listbindings($event) {
    return managebindings('list',$event);
}

function managebindings() {
    static $bind = array();
    $arr = func_get_args();
    $proto = $arr[0];
    $event = $arr[1];
    if (func_num_args()>2) {
	$func = $arr[2];
    }
    switch ($proto) {
	case 'set':
	    if (!array_key_exists($event,$bind)) {
		$bind[$event][] = $func;
	    } else {
		if (!in_array($func,$bind[$event])) {
		    $bind[$event][] = $func;
		}
	    }
	    break;
	case 'unset':
	    if (array_key_exists($event,$bind)) {
		foreach($bind[$event] as $key => $value) {
		    if ($value == $func) {
			array_splice($bind[$event],$key,1);
		    }
		}
	    }
	    break;
	case 'list':
	    if  (array_key_exists($event,$bind)) {
		return $bind[$event];
	    } else {
		return array();
	    }
	    break;
    }
}

bind("changenick","NICK");
function changenick($user,$newnick) {
    if (get_nick($user) == bnick()) {
	bnick($newnick);
    }
}

function bnick() {
    global $bnick;
    if (func_num_args()) {
	$bnick = func_get_arg(0);
	ircwrite("raw", "NICK $bnick");
    }
    return $bnick;
}
						
function is_admin($user) {
    global $badmins;
    return in_array(get_nick($user),$badmins);
}

function randomname($namelength) {
    $namechars = 'abcdefghijklmnopqrstuvwxyz';
    $vouel = 'aeiou';
    $name = "";
    for ($index=1;$index<=$namelength;$index++) {
        if ($index % 3 == 0) {
            $randomnumber = rand(1,strlen($vouel));
            $name .= substr($vouel,$randomnumber-1,1);
         } else {
          $randomnumber = rand(1,strlen($namechars));
          $name .= substr($namechars,$randomnumber-1,1);
         }
    }
    return $name;
}

function rand_name() {
    return ucfirst(randomname(rand(5,8)));
}

function irclog($logproto,$logdata) {
    $irclog = fopen("phpbot.log","a+"); // opening log file
    fwrite($irclog,"$logproto - $logdata\n");
    fclose($irclog);
}

function ircwrite($proto,$ircdata) {
    global $socket;
    $ircdata .= "\r\n"; // auto putting end-signs at every line
    switch ($proto) {
        case "ctcp": // client-to-client protocol
            socket_write($socket,$ircdata);
            break;
        case "msg": // privmsg
            socket_write($socket,"PRIVMSG $ircdata");
            break;
        case "raw": // raw message
            socket_write($socket," $ircdata");
            break;
        default: // everything else
            irclog($proto,$ircdata);
            break;
    }
}

function dump($ircdata) {
    global $socket;
    socket_write($socket,$ircdata."\r\n");
}


echo "exiting\n";

?>
